Imu test - test multiple IMUs and save data

martinaxgloria opened the pull request:
This PR adds the possibility to run the test for multiple IMUs at a time to check the Euler angles values. 

These changes were tested both in simulation and on the real robot (iCubGenova11).

pattacini commented:
Changed base branch to `devel`.

Nicogene commented on src/imu/imu.h:
```diff
@@ -34,11 +34,10 @@
-* | mean_error         | double             | -     | -             | Yes      | The tolerance on the mean of the error. | |
+* | remoteControlBoards| vector of string   | -     | -             | Yes      | The list of the controlboards to open. | e.g. ("torso", "head") |
+* | axesNames          | vector of string   | -     | -             | Yes      | The list of the controlled joints. | e.g. ("torso_pitch", "torso_roll", "torso_yaw", "neck_pitch", "neck_roll", "neck_yaw") |
+* | testList           | vector of string   | -     | -             | Yes      | The list of the sensors to be tested | e.g. ("head_imu_0", "l_arm_ft") |
```
Nitpick: in order to make more clear that in that list the user has to specify the sensor names maybe `sensorsList` is more appropriate?

martinaxgloria replied:
You're right! Done in https://github.com/robotology/icub-tests/pull/69/commits/9f3fc77cd85aa5f9bcb52eb65b8b53e0a594148b

Nicogene reviewed (approved):
Excellent work!

Just a comment not blocking the PR!

martinaxgloria commented:
In https://github.com/robotology/icub-tests/pull/69/commits/46d8d1c47f0601564324a1492448ad19e5d71564 I integrated the `robometry` framework to save relevant measurements (i.e. `expected` and `measured` orientation data, joints `positions` and `velocities`, `error`, and so on) for future inspections.

cc @Nicogene @traversaro

martinaxgloria commented:
Yes, thanks @Nicogene, I'll do it

martinaxgloria commented:
In https://github.com/robotology/icub-tests/pull/69/commits/d35471a404daddd71fe62e6f4c6fd1e018e73154, I implemented the possibility of testing all the available IMUs by passing to the `sensorsList` parameter an empty list or `all`. Moreover, I added a bash that contains some `ctpService` calls to set a sequence of movements of the interested joints (left_arm, right_arm and neck ones) during which the sensors were tested. 

cc @Nicogene @traversaro

martinaxgloria commented:
Tested on iCubGenova11:

https://github.com/robotology/icub-tests/assets/114698424/e548e6ef-9812-430b-9f70-e2b55852ed91

Since https://github.com/robotology/icub-models-generator/issues/233, the test result is:

![fail](https://github.com/robotology/icub-tests/assets/114698424/6bf09955-e75e-41c1-8b10-8a02509fc59e)


cc @Nicogene @traversaro

traversaro commented:
Great!

Nicogene commented on src/imu/imu.cpp:
```diff
@@ -22,89 +25,109 @@ bool Imu::setup(yarp::os::Property& property)
-    frameName = property.find("frame").asString(); // frame on which the sensor is attached
-    sensorName = property.find("sensor").asString(); // sensor name within urdf
-    errorMean = property.find("mean_error").asFloat64(); //error mean
+    errorMean = property.find("meanError").asFloat64(); //error mean
```
This is actually `maxError` am I right?

Nicogene commented on src/imu/imu.cpp:
```diff
@@ -22,89 +25,109 @@ bool Imu::setup(yarp::os::Property& property)
+        ROBOTTESTINGFRAMEWORK_TEST_REPORT("Testing all the IMUs available...");
+        sensorsNamesList.addString("l_arm_ft");
+        sensorsNamesList.addString("r_arm_ft");
+        sensorsNamesList.addString("head_imu_0");
```
Why `all` add only those sensors? Maybe it is better to ask idyntree the list of all sensors, because those names can change

traversaro replied:
> If I ask iDynTree for the list of the sensors of type `iDynTree::ACCELEROMETER`, it returns 62 sensors because it includes also the mtb
> 
> cc @traversaro

It would be great if we could test all the mtb4 sensors! Do you think there is anything preventing us to do so? I guess for known issues we could have an additional parameter to "exclude" sensors instead of adding them.

martinaxgloria replied:
Since so far the configuration files to read from the mtb are not updated (see for example https://github.com/robotology/icub-models/blob/devel/iCub/conf/MTB/gazebo_icub_left_arm_mtb.ini) and we should add all the 62 sensors to the `alljoints-inertials`, we could do a workaround: if we want to test `all` the sensors _available_, we could use another `IOrientationSensors` and make the `multipleanalogsensorsclient` do the view on this, so that we could retrieve only the orientation sensors that are listed in the `alljoints-inertials.xml` and exclude the mtb for the time being 

cc @Nicogene @traversaro

Nicogene commented on src/imu/imu.cpp:
```diff
@@ -195,57 +283,24 @@ bool Imu::sendData(iDynTree::Vector3 expectedValues, iDynTree::Vector3 imuSignal
 }
 
-bool Imu::moveJoint(int ax, double pos)
+bool Imu::setupTelemetry()
```
Niptick: robometry :D

Nicogene commented on src/imu/imu.h:
```diff
@@ -34,11 +34,10 @@
+* | remoteControlBoards| vector of string   | -     | -             | Yes      | The list of the controlboards to open. | e.g. ("torso", "head") |
+* | axesNames          | vector of string   | -     | -             | Yes      | The list of the controlled joints. | e.g. ("torso_pitch", "torso_roll", "torso_yaw", "neck_pitch", "neck_roll", "neck_yaw") |
+* | sensorsList        | vector of string   | -     | -             | Yes      | The list of the sensors to be tested | e.g. ("head_imu_0", "l_arm_ft") or ("all)|
+* | meanError          | double             | -     | -             | Yes      | The tolerance on the error. | |
```
Same as above

Nicogene reviewed (changes requested):
Great work @martinaxgloria!

Some small points to be addressed, but we have almost landed 🛬

Have you forgot to commit the `move.sh` script?

Nicogene reviewed (changes requested):
Ah I forgot, now that the CI is conda based should be easy to enable the compilation of this test in the CI adding also the additional dependencies

martinaxgloria commented:
> Ah I forgot, now that the CI is conda based should be easy to enable the compilation of this test in the CI adding also the additional dependencies

Done here https://github.com/robotology/icub-tests/pull/69/commits/936c4e37ee7570e70b4a4a75900d0ea61895f8f6 and https://github.com/robotology/icub-tests/pull/69/commits/513a57a8514ea529e43d954ee4ea15b6d9ab8d8a

cc @Nicogene
